File: src/main/java/eu/mshade/mwork/binarytag/BinaryTagType.java
package eu.mshade.mwork.binarytag;

import eu.mshade.mwork.binarytag.entity.*;

import java.util.HashMap;
import java.util.Map;


public enum BinaryTagType {

    END(EndBinaryTag.class, 0),
    BYTE(ByteBinaryTag.class,1),
    SHORT(ShortBinaryTag.class, 2),
    INTEGER(IntegerBinaryTag.class,3),
    LONG(LongBinaryTag.class, 4),
    FLOAT(FloatBinaryTag.class, 5),
    DOUBLE(DoubleBinaryTag.class, 6),
    BYTE_ARRAY(ByteArrayBinaryTag.class, 7),
    STRING(StringBinaryTag.class, 8),
    LIST(ListBinaryTag.class, 9),
    COMPOUND(CompoundBinaryTag.class, 10),
    INTEGER_ARRAY(IntegerArrayBinaryTag.class,11),
    LONG_ARRAY(LongArrayBinaryTag.class,12),

    BOOLEAN(BooleanBinaryTag.class, 13),

    SHADE_BYTE_ARRAY(ShadeByteArrayBinaryTag.class, 14),
    SHADE_LIST(ShadeListBinaryTag.class, 15),
    SHADE_COMPOUND(ShadeCompoundBinaryTag.class, 16),
    SHADE_INTEGER_ARRAY(ShadeIntegerArrayBinaryTag.class, 17),
    SHADE_LONG_ARRAY(ShadeLongArrayBinaryTag.class, 18),

    ZSTD_BYTE_ARRAY(ZstdByteArrayBinaryTag.class, 19),
    ZSTD_LIST(ZstdListBinaryTag.class, 20),
    ZSTD_COMPOUND(ZstdCompoundBinaryTag.class, 21),
    ZSTD_INTEGER_ARRAY(ZstdIntegerArrayBinaryTag.class, 22),
    ZSTD_LONG_ARRAY(ZstdLongArrayBinaryTag.class, 23);



    private static final Map<Integer, BinaryTagType> BY_ID = new HashMap<>();
    private static final Map<Class<? extends BinaryTag<?>>, BinaryTagType> BY_CLASS = new HashMap<>();

    static {
        for (BinaryTagType binaryTagType : BinaryTagType.values()) {
            BY_ID.put(binaryTagType.getType(), binaryTagType);
            BY_CLASS.put(binaryTagType.getClazz(), binaryTagType);
        }
    }

    private final Class<? extends BinaryTag<?>> aClass;
    private final int id;

    BinaryTagType(Class<? extends BinaryTag<?>> aClass, int id) {
        this.aClass = aClass;
        this.id = id;
    }

    public int getType() {
        return id;
    }

    public Class<? extends BinaryTag<?>> getClazz() {
        return aClass;
    }

    public static BinaryTagType getTagTypeById(int id){
        return BY_ID.get(id);
    }

    public static BinaryTagType getTagTypeByClass(Class<? extends BinaryTag<?>> aClass){
        return BY_CLASS.get(aClass);
    }
}

File: src/main/java/eu/mshade/mwork/binarytag/BinaryTagType.kt
package eu.mshade.mwork.binarytag

import eu.mshade.mwork.binarytag.entity.*

object BinaryTagType {

    @JvmField val END = BinaryTagKey.from(0, "End",EndBinaryTag::class)
    @JvmField val BYTE = BinaryTagKey.from(1, "Byte", ByteBinaryTag::class)
    @JvmField val SHORT = BinaryTagKey.from(2, "Short", ShortBinaryTag::class)
    @JvmField val INT = BinaryTagKey.from(3, "Int", IntBinaryTag::class)
    @JvmField val LONG = BinaryTagKey.from(4, "Long", LongBinaryTag::class)
    @JvmField val FLOAT = BinaryTagKey.from(5, "Float", FloatBinaryTag::class)
    @JvmField val DOUBLE = BinaryTagKey.from(6, "Double", DoubleBinaryTag::class)
    @JvmField val BYTE_ARRAY = BinaryTagKey.from(7, "ByteArray", ByteArrayBinaryTag::class)
    @JvmField val STRING = BinaryTagKey.from(8, "String", StringBinaryTag::class)
    @JvmField val LIST = BinaryTagKey.from(9, "List", ListBinaryTag::class)
    @JvmField val COMPOUND = BinaryTagKey.from(10, "Compound", CompoundBinaryTag::class)
    @JvmField val INT_ARRAY = BinaryTagKey.from(11, "IntArray", IntArrayBinaryTag::class)
    @JvmField val LONG_ARRAY = BinaryTagKey.from(12, "LongArray", LongArrayBinaryTag::class)

    @JvmField val BOOLEAN = BinaryTagKey.from(13, "Boolean", BooleanBinaryTag::class)

    @JvmField val ZSTD_BYTE_ARRAY = BinaryTagKey.from(14, "ZstdByteArray", ZstdByteArrayBinaryTag::class)
    @JvmField val ZSTD_LIST = BinaryTagKey.from(15, "ZstdList", ZstdListBinaryTag::class)
    @JvmField val ZSTD_COMPOUND = BinaryTagKey.from(16, "ZstdCompound", ZstdCompoundBinaryTag::class)
    @JvmField val ZSTD_INT_ARRAY = BinaryTagKey.from(17, "ZstdIntArray", ZstdIntArrayBinaryTag::class)
    @JvmField val ZSTD_LONG_ARRAY = BinaryTagKey.from(18, "ZstdLongArray", ZstdLongArrayBinaryTag::class)

    @JvmField val DEFLATE_BYTE_ARRAY = BinaryTagKey.from(19, "DeflateByteArray", DeflateByteArrayBinaryTag::class)
    @JvmField val DEFLATE_LIST = BinaryTagKey.from(20, "DeflateList", DeflateListBinaryTag::class)
    @JvmField val DEFLATE_COMPOUND = BinaryTagKey.from(21, "DeflateCompound", DeflateCompoundBinaryTag::class)
    @JvmField val DEFLATE_INT_ARRAY = BinaryTagKey.from(22, "DeflateIntArray", DeflateIntArrayBinaryTag::class)
    @JvmField val DEFLATE_LONG_ARRAY = BinaryTagKey.from(23, "DeflateLongArray", DeflateLongArrayBinaryTag::class)


    private val binaryTagTypeById = mutableMapOf<Int, BinaryTagKey>()

    init {
        BinaryTagType::class.java.declaredFields.forEach { field ->
            val obj = field.get(null)
            if (obj is BinaryTagKey) {
                binaryTagTypeById[obj.getIdentifier()] = obj
            }
        }
    }

    fun getTagTypeById(id: Int): BinaryTagKey {
        return binaryTagTypeById[id] ?: throw IllegalArgumentException("Unknown BinaryTagTypeKey with id $id")
    }

    fun getTagTypeById(id: Byte): BinaryTagKey = getTagTypeById(id.toInt())

    fun registerTagType(binaryTagKey: BinaryTagKey) {
        binaryTagTypeById[binaryTagKey.getIdentifier()] = binaryTagKey
    }



}